rdflib/plugins/sparql/evaluate.py

```diff
@@ -381,18 +381,22 @@ def _yieldBindingsFromServiceCallResult(ctx: QueryContext, r, variables):
     res_dict: Dict[Variable, Identifier] = {}
     for var in variables:
         if var in r and r[var]:
-            if r[var]["type"] == "uri":
-                res_dict[Variable(var)] = URIRef(r[var]["value"])
-            elif r[var]["type"] == "bnode":
-                res_dict[Variable(var)] = BNode(r[var]["value"])
-            elif r[var]["type"] == "literal" and "datatype" in r[var]:
+            d = r[var]
+            t = d["type"]
+            if t == "uri":
+                res_dict[Variable(var)] = URIRef(d["value"])
+            elif t == "literal":
                 res_dict[Variable(var)] = Literal(
-                    r[var]["value"], datatype=r[var]["datatype"]
+                    d["value"], datatype=d.get("datatype"), lang=d.get("xml:lang")
                 )
-            elif r[var]["type"] == "literal" and "xml:lang" in r[var]:
+            elif t == "typed-literal":
```

Comment:
I'm don't think `typed-literal` is a valid type:

https://www.w3.org/TR/sparql11-results-json/#select-encode-terms

![image](https://user-images.githubusercontent.com/4481387/168140718-daf3e8fb-2dc7-4532-98ba-16f1d2d20e93.png)

The reason it is here is because of https://www.w3.org/TR/2006/NOTE-rdf-sparql-json-res-20061004/#variable-binding-results AFAIK - so best to add a comment here.

Reply:
@aucampia To my opinion both are valid comments. I reused the logic from the JSON results parser: https://github.com/RDFLib/rdflib/blob/6f2c11cd2c549d6410f9a1c948ab3a8dbf77ca00/rdflib/plugins/sparql/results/jsonresults.py#L89-L107
Maybe it is a good idea to apply the same changes there too.

Reply:
I will do that in a separate PR, trying to figure out how to reproduce a BNode in response.

Reply:
I also considered adding BNode to the test, but could not come up with a good way to do it, so it would be nice if you can figure this out!

Reply:
I kind of gave up, I will go with using the mock HTTP server that is now in master and check it in a separate test.

Reply:
Also I will add this comment shortly, will tell you when done but will get everything to a state where I approve tonight still.

Reply:
Added mock based tests now.